crates/uv-resolver/src/lock.rs

```diff
@@ -588,62 +682,203 @@ impl Lock {
+        tags: &Tags,
+        database: &DistributionDatabase<'_, Context>,
+    ) -> Result<bool, LockError> {
+        let mut queue: VecDeque<&Package> = VecDeque::new();
+        let mut seen = FxHashSet::default();
+
+        // Validate that the lockfile was generated with the same root members.
+        {
+            let expected = members;
+            let actual = &self.members;
+            if expected != actual {
+                debug!(
+                    "Mismatched members:\n  expected: {:?}\n  found: {:?}",
+                    expected, actual
+                );
+                return Ok(false);
             }
+        }
 
-            match package.id.source {
-                Source::Registry(..) | Source::Git(..) => {}
-                // Skip local and direct URL dependencies, as their metadata may have been mutated
-                // without a version change.
-                Source::Path(..)
-                | Source::Directory(..)
-                | Source::Editable(..)
-                | Source::Direct(..) => continue,
+        // Validate that the lockfile was generated with the same constraints.
+        {
+            let expected: Vec<_> = constraints
+                .iter()
+                .cloned()
+                .map(|requirement| normalize_requirement(requirement, workspace))
+                .collect();
+            let actual: Vec<_> = self
+                .constraints
+                .iter()
+                .cloned()
+                .map(|requirement| normalize_requirement(requirement, workspace))
+                .collect();
+            if expected != actual {
+                debug!(
+                    "Mismatched constraints:\n  expected: {:?}\n  found: {:?}",
+                    expected, actual
+                );
+                return Ok(false);
             }
+        }
 
-            // Add registry distributions to the package index.
-            if let Some(prioritized_dist) = package.to_prioritized_dist(install_path)? {
-                packages
-                    .entry(package.name().clone())
-                    .or_default()
-                    .insert(package.id.version.clone(), prioritized_dist);
+        // Validate that the lockfile was generated with the same overrides.
+        {
+            let expected: Vec<_> = overrides
+                .iter()
+                .cloned()
+                .map(|requirement| normalize_requirement(requirement, workspace))
+                .collect();
+            let actual: Vec<_> = self
+                .overrides
+                .iter()
+                .cloned()
+                .map(|requirement| normalize_requirement(requirement, workspace))
+                .collect();
+            if expected != actual {
+                debug!(
+                    "Mismatched overrides:\n  expected: {:?}\n  found: {:?}",
+                    expected, actual
+                );
+                return Ok(false);
             }
+        }
```

Comment:
I think these are all fine, but I think will result in reporting "not satisfied" in cases where the lock file does actually satisfy the requirements. But I think this is something we can improve in follow-up PRs. Specifically, even if the dependency specification in the `pyproject.toml` changes, the version that's in the lock file can still satisfy it such that no actual updates are needed. But I think this implementation will say an update is needed if _any_ change occurs.

Reply:
Yeah this may actually be too conservative... I mostly care about this capability to ensure that `uv run` is really fast / cheap, so I'm ok with having to do a resolution if the user edits the dependencies.